modules/core/keeper/msg_server.go

```diff
@@ -1084,12 +1084,25 @@ func (k Keeper) ChannelUpgradeCancel(goCtx context.Context, msg *channeltypes.Ms
 		return nil, errorsmod.Wrap(err, "channel upgrade cancel failed")
 	}
 
+	// grab upgrade here since it will be deleted in WriteUpgradeCancelChannel
+	upgrade, found := k.ChannelKeeper.GetUpgrade(ctx, msg.PortId, msg.ChannelId)
+	if !found {
+		return nil, errorsmod.Wrapf(channeltypes.ErrUpgradeNotFound, "failed to retrieve channel upgrade: port ID (%s) channel ID (%s)", msg.PortId, msg.ChannelId)
+	}
+
 	k.ChannelKeeper.WriteUpgradeCancelChannel(ctx, msg.PortId, msg.ChannelId, msg.ErrorReceipt.Sequence)
 
 	cbs.OnChanUpgradeRestore(ctx, msg.PortId, msg.ChannelId)
 
 	ctx.Logger().Info("channel upgrade cancel succeeded", "port-id", msg.PortId, "channel-id", msg.ChannelId)
 
+	// get channel here again to get latest state after write
+	channel, found = k.ChannelKeeper.GetChannel(ctx, msg.PortId, msg.ChannelId)
+	if !found {
+		return nil, errorsmod.Wrapf(channeltypes.ErrChannelNotFound, "port ID (%s) channel ID (%s)", msg.PortId, msg.ChannelId)
+	}
+	keeper.EmitChannelUpgradeCancelEvent(ctx, msg.PortId, msg.ChannelId, channel, upgrade)
```

Comment:
Nice that you moved this here. I guess we can then [remove it from `WriteUpgradeCancelChannel`](https://github.com/cosmos/ibc-go/blob/bb42c9a14f274c903773211fa072365b3d0ad465/modules/core/04-channel/keeper/upgrade.go#L683C2-L683C31)!

Reply:
aha, yes, i think that was either missed or maybe the merge left the artifact as i do remember there was a PR to remove all of those at some point... i'll take it out now!